game-core/src/main/java/games/strategy/triplea/ai/pro/util/ProTerritoryValueUtils.java

```diff
@@ -177,19 +176,31 @@ public static double findTerritoryAttackValue(
     return territoryValueMap;
   }
 
-  private static int findMaxLandMassSize(final GamePlayer player) {
-    int maxLandMassSize = 1;
+  protected static int findMaxLandMassSize(final GamePlayer player) {
     final GameData data = player.getData();
+    final Predicate<Territory> cond = ProMatches.territoryCanPotentiallyMoveLandUnits(player, data);
+    class LandMassFinder extends BreadthFirstSearch {
+      int landMassSize = 0;
+
+      LandMassFinder(final Territory startTerritory) {
+        super(startTerritory, cond);
+      }
+
+      @Override
+      public void visit(final Territory territory) {
+        landMassSize++;
+      }
+    }
+
+    int maxLandMassSize = 1;
+    final var fullVisited = new HashSet<Territory>();
     for (final Territory t : data.getMap().getTerritories()) {
-      if (!t.isWater()) {
-        final int landMassSize =
-            1
-                + data.getMap()
-                    .getNeighbors(
-                        t, 6, ProMatches.territoryCanPotentiallyMoveLandUnits(player, data))
-                    .size();
-        if (landMassSize > maxLandMassSize) {
-          maxLandMassSize = landMassSize;
+      if (!t.isWater() && !fullVisited.contains(t)) {
+        final LandMassFinder finder = new LandMassFinder(t);
+        finder.search();
+        fullVisited.addAll(finder.getVisited());
```

Comment:
What do you think of having `finder.seach()` returning the `visited` list? This would avoid the void method call, and avoid the user of `finder` from knowing about the implementation detail that `finder.search` sets a value that can be returned by `getVisited` (arguably better/encapsulation)

Reply:
Another option is to not expose visited() at all and have the caller's visit() method override keep track of this case.

Reply:
That also sounds like a reasonably alternative. 

If I understand right, there would be no side-effects as the data state would be self-contained to the BFS implementation, do I have that right? If that can be done cleanly, it sounds good.

Reply:
Done.